Test fixes

djmitche opened the pull request:
This should make the tests a little more reliable.. one hopes.  We'll see how the PR tests go.

djmitche committed:
use tdd hooks where possible, clean up carefully

djmitche committed:
don't use async test functions with done callback

Mocha spots the `done` parameter and ignores the returned Promise,
leading to hangs when async exceptions are thrown.

djmitche committed:
support errors from async event handlers

djmitche committed:
skip pulse publishing for testing handlers

imbstack commented on src/handlers.js:
```diff
@@ -68,8 +68,16 @@ class Handlers {
-    this.statusListener.on('message', this.monitor.timedHandler('statuslistener', statusHandler.bind(this)));
+    const callHandler = (name, handler) => message => {
+      handler.call(this, message).catch(err => {
+        debug(`error calling ${name} handler: ${err}`);
```
Maybe we should say "error reported to sentry" or something along those lines?

djmitche replied:
Good idea

imbstack commented on test/api_test.js:
```diff
@@ -46,55 +46,35 @@ suite('api', () => {
-    } catch (e) {
-      done(e);
-    }
+    assert.equal(builds.builds.length, 4);
```
iirc, if line 50 `await helper.github.builds();` throws an exception, this test will hang rather than complete now? I might be mis-remembering though.

djmitche replied:
No, because it's an async function, mocha will wait for the Promise, and an exception in an async function will reject the Promise.

djmitche committed:
Don't even connect to pulse, test job and status handlers separately

This avoids the need to claim and complete any tasks, so there's no need
for a pulse listener.

djmitche committed:
await the handler completion, instead of polling

polling is evil.

djmitche commented:
Ok, I'm done playing with this.  Since it was intermittent, I'm not sure (a) which of these fixes stopped the intermittents or (b) if any of them did.  But hopefully fixing async stuff, taking pulse out of the equation, taking queue operations out of the equation, and avoiding polling will be sufficient.
